### PR TITLE
Tests for Kathirvalavakumar Subavathi Initialisation

### PR DESCRIPTION
**Pull request by chvsp**
I have written the  tests for Iris Dataset, Non-Linear Function Approximation tasks using Kathirvalavakumar Subavathi Initialisation method, as outlined in the paper.

### CODE REVIEW

**Comment by chvsp** [comment]
@zoq , Kindly have a look at this PR.

File: src/mlpack/methods/ann/ffn.hpp [hunk 1]
```diff
@@ -25,6 +25,7 @@
 #include <mlpack/methods/ann/layer/layer_types.hpp>
 #include <mlpack/methods/ann/init_rules/random_init.hpp>
 #include <mlpack/methods/ann/init_rules/nguyen_widrow_init.hpp>
+#include <mlpack/methods/ann/init_rules/kathirvalavakumar_subavathi_init.hpp>
```
Comment (zoq) [inline src/mlpack/methods/ann/ffn.hpp]: I think as long as we don't use the kathirvalavakumar subavathi initialization method in the FFN class, we don' have to include the header file, I guess the same applies for the nguyen widrow method.

File: src/mlpack/tests/ksinit_test.cpp [hunk 2]
```diff
@@ -0,0 +1,315 @@
+/*
+  @file ksinit_test.cpp
+  @author Praveen Ch
+  
+
+  Tests the working of Kathirvalavakumar Subavathi Initialization for a
```
Comment (zoq) [inline src/mlpack/tests/ksinit_test.cpp]: Can you also add the licence:

```
/**
 * @file ksinit_test.cpp
 * @author Praveen Ch
 *
 * Tests the working of Kathirvalavakumar Subavathi Initialization for a 
 * Feed forward neural network.
 *
 * mlpack is free software; you may redistribute it and/or modify it under the
 * terms of the 3-clause BSD license.  You should have received a copy of the
 * 3-clause BSD license along with mlpack.  If not, see
 * http://www.opensource.org/licenses/BSD-3-Clause for more information.
 */
```

File: src/mlpack/tests/ksinit_test.cpp [hunk 3]
```diff
@@ -0,0 +1,315 @@
...
+
+  return avgError;
+}
+
+
+/*Test case for the Iris Dataset */
+
+BOOST_AUTO_TEST_CASE(IrisDataset)
+{
+  arma::arma_rng::set_seed_random();
```
Comment (zoq) [inline src/mlpack/tests/ksinit_test.cpp]: We set the random seed for each build. Enabling us to reuse the seed if something breaks to e.g. track down an error that only occurs in certain situations. If you reset the seed for the test case, we can't do the same for this test suite.

File: src/mlpack/tests/ksinit_test.cpp [hunk 4]
```diff
@@ -0,0 +1,315 @@
...
+                         MatType& testData,
+                         MatType& testLabels,
+                         const size_t hiddenLayerSize,
+                         const size_t maxEpochs,
+                         const size_t outputSize,
+                         double& trainError,
+                         double& testError)
+{
+  /*
+  @param trainError mean squared error of predictions on training data.
```
Comment (zoq) [inline src/mlpack/tests/ksinit_test.cpp]: Minor style issue, if you comment the parameters for the test, put the documentation over the function definition, that way  Doxygen can generate the documentation.

File: src/mlpack/tests/ksinit_test.cpp [hunk 5]
```diff
@@ -0,0 +1,315 @@
...
+/* Error is a structure which has the MSE of Training data and MSE of 
+   Validation data respectively. This structure is returned to the parent test 
+   case by CrossValidation function.
+
+*/
+struct Error
+{
+  double trainError;
+  double validationError;
+};
```
Comment (zoq) [inline src/mlpack/tests/ksinit_test.cpp]: I guess, just to be consistent with the rest of the code style used in your test, I would go with a reference for ``trainError`` and ``validationError`` as you have done for the BuildVanillaNetwork function. What do you think?

File: src/mlpack/tests/ksinit_test.cpp [hunk 6]
```diff
@@ -0,0 +1,315 @@
...
+*/
+
+Error CrossValidation(arma::mat& trainData, arma::mat& trainLabels, size_t k,
+                      const size_t hiddenLayerSize, const size_t maxEpochs)
+{
+
+/*
+  @params trainData The data available for training.
+  @params trainLabels The labels corresponding to the training data.
+  @params k The parameter k in k-fold cross validation.
```
Comment (zoq) [inline src/mlpack/tests/ksinit_test.cpp]: Please put the function documentation above the function definition.

File: src/mlpack/tests/ksinit_test.cpp [hunk 7]
```diff
@@ -0,0 +1,315 @@
...
+    arma::mat trainData = dataset.submat(0, 0, dataset.n_rows - 1 - numLabels, 
+                                         dataset.n_cols - 1);
+    arma::mat trainLabels = dataset.submat(dataset.n_rows - numLabels, 0, 
+                                           dataset.n_rows - 1, 
+                                           dataset.n_cols - 1); 
+    Error e = CrossValidation(trainData, trainLabels, 10, hiddenLayerSize, 
+                              maxEpochs);
+
+    // The check is to prevent nan values from affecting the result.
+    if (arma::is_finite(e.trainError) && arma::is_finite(e.validationError))
```
Comment (zoq) [inline src/mlpack/tests/ksinit_test.cpp]: Have you encountered any problems in that direction?

File: src/mlpack/tests/ksinit_test.cpp [hunk 8]
```diff
@@ -0,0 +1,315 @@
...
+  Error avgError = AvgCrossValidation(dataset, 1, 10, 3, 52); //Runs CV 10 times
+
+  BOOST_REQUIRE_LE(avgError.trainError, trainErrorThreshold);
+  BOOST_REQUIRE_LE(avgError.validationError, validationErrorThreshold);
+
+}
+
+// Test case for the Non Linear Function Approximation Problem.
+
+BOOST_AUTO_TEST_CASE(NonLinearFunctionApproximation)
```
Comment (zoq) [inline src/mlpack/tests/ksinit_test.cpp]: I have completely forgotten the non linear function approximation task, looks nice. Thanks for writing another test case :+1:

File: src/mlpack/tests/ksinit_test.cpp (hunk 5 above)
Comment (chvsp) [inline src/mlpack/tests/ksinit_test.cpp]: I had an issue with naming the variables. There were so many trainErrors, validationErrors that I found creating a structure helpful as it reduces my variable names by half. If you could suggest some way to handle this, I can change it. Or else I can modify BuildVanillaNetwork function to return Error. Kindly advise.

File: src/mlpack/tests/ksinit_test.cpp (hunk 2 above)
Comment (chvsp) [inline src/mlpack/tests/ksinit_test.cpp]: Added the license.

**Comment by chvsp** [comment]
@zoq Kindly review, I have solved all the issues. The build sometimes fails albeit very rarely. Should I increase the threshold even further, or is it ok?

**Comment by zoq** [comment]
Since it's not guaranteed that the method will converge in the specific situation it's safe if we run the test multiple times and if this works e.g. 1 of 5 times it's fine. What do you think?

**Comment by chvsp** [comment]
This is a good idea but I think running it for 5 iterations everytime will be an overkill. Can we break it once it has passed the criteria? I think it would be equally safe.

**Comment by zoq** [comment]
Yes, no need to continue once it passed.

**Comment by chvsp** [comment]
Ok. Will make the modifications and push the commits.

**Comment by chvsp** [comment]
@zoq I have made the necessary changes. My tests have passed the build. GMMTest has failed. Kindly review the code. Thanks :)

**Comment by zoq** [comment]
Thanks for the contribution, I made some minor formatting fixes and clarifications in 847b5ac0ffe86fc9.

**Comment by chvsp** [comment]
@zoq Thank you for reviewing the code. I will keep the formatting errors in mind. :)

**Pull request merged**